controllers/metal3.io/baremetalhost_controller.go

```diff
@@ -932,7 +932,7 @@ func (r *BareMetalHostReconciler) manageHostPower(prov provisioner.Provisioner,
 	if desiredPowerOnState {
 		provResult, err = prov.PowerOn()
 	} else {
-		provResult, err = prov.PowerOff(desiredRebootMode)
+		provResult, err = prov.PowerOff(desiredRebootMode, info.host.Status.ErrorType == metal3v1alpha1.PowerManagementError)
```

Comment:
Under what circumstances can we expect host.Status.ErrorType to be set in general and specifically to  metal3v1alpha1.PowerManagementError? This fix depends on Ironic setting host.Status.ErrorType to metal3v1alpha1.PowerManagementError when there is an issue while soft or hard powering off a host and that the ironicNode.LastError is also set during these failure scenarios.

Reply:
It's set on line 942 and cleared on line 956. (There's actually a bug, because it should also be cleared around line 966.) So it will be set whenever an error is reported and cleared whenever we manage to do something (anything) without immediately failing.
This is the right logic for a force flag to ignore the LastError in ironic and retry. Each time a retry is started, the flag will be cleared so if it fails we'll report the error again.
However, I think it's probably the wrong logic for forcing us into the hard power off mode (which is how this param is currently being used). IIRC we discussed implementing it so that we only try a soft power off once and any retries go straight to pulling the plug. For that we want to check for a non-zero ErrorCount, because the ErrorCount doesn't get cleared until the power management operation actually succeeds.

Reply:
Thanks! That was my concern. Updated to use ErrorCount to determine the force flag. Do you want me to take care of the other bug regarding clearing the ErrorType here?